### ISSUE TITLE
Running PrisM

### ISSUE THREAD

**Issue by sadger** [anaarmas/PrisM#1]
Hi i'm trying to get a working version of PrisM. I edited the pom.xml as required. However, when I try and compile it with `mvn install` it generates a jar file successfully. I include the built jar file into my eclipse project as usual but gives an error about missing classes.

```
Caused by: java.lang.ClassNotFoundException: uk.ac.ox.cs.JRDFox.JRDFStoreException
    at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
    at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
```

 I suspect the maven built is not producing a self contained jar.
Do you have a precompiled linux jar that I can use or more instructions how to build it correctly?

Thanks.

**Comment by anaphylactic** [comment]
Hi,
Thanks for downloading PrisM and sorry that you had problems trying to use it! I believe I have now fixed the issue, if you update the project and follow the new instructions in the README file it should work correctly - feel free to let me know if it doesn't!
Cheers,
Ana

**Comment by sadger** [comment]
Great. Seems to be working now. Thanks for taking the time to make those changes. I'll see if I can get some modules extracted now.

**Comment by anaphylactic** [comment]
No problem at all! I hope you will find the tool useful ;)

**Comment by sadger** [comment]
Hi, I have managed to get some module extraction working with PrisM, I am particularly interesting in the model-inseparability relation and seeing how good an approximation PrisM produces. I'm one of the authors from " Lower and Upper Approximations for Depleting Modules of Description Logic Ontologies" in which we compared STAR (\bot\top*) and AMEX (and a combination of both modules) to a lower approximation to estimate how close the modules were to the ideal minimal ones. 

One problem I am having with PrisM is extracting modules for several signatures, for a single extraction it works correctly producing a module.

Is it possible to extract multiple modules with a single instance of PrisM without unloading and reloading the ontology? I am getting an  `OWLOntologyAlreadyExistsException` when trying to do the following.

``` java
PrisM extractor = new PrisM(ont, InseparabilityRelation.MODEL_INSEPARABILITY);
extractor.extract(signature1,false)
extractor.extract(signature2,false)
extractor.extract(signature3,false)
...

```

Is there a way to make it work to extract multiple signatures, maybe I am going about it incorrectly?

Thanks again,

William

**Comment by anaphylactic** [comment]
Hi William,

I am familiar with your work, of course! :)
I think I have fixed that bad behaviour now, if you update the project and rebuild the jar it should work as you need it to. Do let me know if you encounter any other problems.
And thanks for your comments, this helps make the system better!

Cheers,
Ana

**Comment by sadger** [comment]
Great! Seems to work well extracting multiple signatures. I'll try testing it on a larger amount of ontologies and see well it performs. I really appreciate you putting the time in to fix these issues I have been having.

Thanks again,

William

**Comment by sadger** [comment]
Sorry, got the same issue again with the  `OWLOntologyAlreadyExistsException`.
It only happens now if I do `extract(signature,false)` when it is `extract(signature,true)` I don't have the issue which is an improvement based on your changes.

**Comment by anaphylactic** [comment]
That constructor should actually be private and only for testing purposes - I've changed that now. 
The first thing that PrisM does is extracting a preliminary syntactic locality based module. That simplifies the work to be done in a cheap way. Calling the extract method with that boolean flag set to false makes this step get skipped, but there is no advantage in that so it doesn't make sense to keep that constructor public. Hope that makes sense!
Ana

**Comment by sadger** [comment]
Oh ok, that makes sense. I can understand pre-processing with the locality based modules we do it in some of our extraction algorithms too. Is the same module produced with and without the preprocessing step? If it was only intended to be used with the locality preprocessing I will do some investigation using that approach, not a problem. I appreciate you putting the time in to make these changes and answer my queries.

William

**Comment by anaphylactic** [comment]
It's not a problem at all, it's very nice to have somebody interested in the work! :)
In ontologies with 'right hand side' disjunctions we may get a smaller module with the preprocessing step; in all other ontologies it doesn't make a difference.

**Comment by sadger** [comment]
Ok great, thanks. I'll give it a go with the preprocessing as you intended then.

**Comment by sadger** [comment]
Hi, I've unfortunately got a couple more issues that I have discovered while testing PrisM.

1) If the ontology does not have a global ontology IRI or document IRI which is quite common depending how the ontology is generated, PrisM will cause a null pointer exception. I can work around this by assigning the ontology one of course.

2) I am getting some strange output for some ontologies and signatures, axioms that look like 
`A ⊑ ∀ r.<internal:def#1>`. I have produced a small ALCQ ontology (10 axioms) and signature for which this problem can be reproduced. For comparative purposes the locality STAR module I extracted contains 1 axiom for this signature whereas the datalog generated module (for the model inseparability relation) contains 11 (one more than the entire ontology), including these weird symbols. Does my input ontology have to be in a certain format for PrisM to work?

Ontology: www.csc.liv.ac.uk/~wgatens/small-example.owl
Signature: www.csc.liv.ac.uk/~wgatens/example-sig

**Comment by anaphylactic** [comment]
I will try to fix the IRI issue ASAP. 
The other issue was already in my todo list, but that may take a little longer. This happens because PrisM works with a normalised version of the input ontology and does not currently undo this normalisation. What it outputs is still inseparable from the original ontology w.r.t. the given signature according to the chosen inseparability relation, but it is not, strictly speaking, a module of the original ontology - rather a module of its normalisation.

**Comment by sadger** [comment]
Ok, I can understand why that would happen if you are normalising first. For comparative purposes I think I will wait until the implementation extracts a module of the original ontology as I cannot accurately compare the sizes of modules, the main metric of interest, at the moment.

Please do not feel the need to fix these things quickly on my behalf. If you just let me know when there is a new version to try I'll happily do some more testing. 

Thanks,

William

**Comment by anaphylactic** [comment]
No worries - all the issues you're mentioning are fair enough and I did need to fix them anyway. And reversing the normalisation really was in my plans too :). I will let you know when I have finished that - thanks for your patience!
Ana

**Comment by sadger** [comment]
Hi Ana, Hope you are well. 

My supervisor brought up an interesting idea, wondering if its feasible.  Could one normalise an ontology to work with PrisM then extract a PrisM/locality/AMEX module from the normalisation and for a signature compare the sizes of such modules. The output may still have these fresh symbols but I would hope they are directly comparable, maybe it is not so simple.

Thanks,

William

**Comment by anaphylactic** [comment]
Hi William,
That is of course possible: if the input ontology is already normalised then there is no further normalisation needed once it is given to PrisM. That is indeed the approach I followed for my own experiments. I you wish I can provide you with a class that will produce for you the same normalisation of the ontology that PrisM would work with. In any case, I'm planning to finish implementing the reversion of the normalisation by the end of October. Depending on your time constraints, you may prefer to wait or to get the class that I mentioned - let me know :)
Best,
Ana

**Comment by sadger** [comment]
Hi Ana, 

If you could provide a class performing the normalisation that would be very helpful to me.
Just to perform some preliminary investigations, once the reversion of normalisation is done I will attempt to do some more extensive comparisons, but no rush on that front.

Thanks,

William

**Comment by anaphylactic** [comment]
Hi William,
The method getNormalisedAxioms in class test/thesisTests/NormaliserViaClauses (now updated) does what you need. Let me know how it goes! :) And I'll keep you posted about the reversion of the normalisation.
Cheers,
Ana

**Comment by sadger** [comment]
Hi Ana, Thanks so much for doing this, will be very useful, I'll try it out later and see how it goes. Thanks again for all your help. William

**Comment by anaphylactic** [comment]
No problem! :D

**Comment by anaphylactic** [comment]
Hey William,
The reversion of the normalisation is now finished. 
Cheers!
Ana

**Comment by sadger** [comment]
Hi Ana, That is excellent news! Thanks for letting me know. I'll give it a go sometime over the next few days and let you know how I find it. 

William

**Comment by sadger** [comment]
Hi Ana, 

Not sure if you are still involved in working on this but I was wondering if you can help me. By default the modules produced by this PrisM implementation **strongly** depleting or just depleting? I'm aware you have an additional setting for your datalog approach which allows strongly depleting modules to be produced, but wondering which it is PrisM is producing, especially for the model-inseparability relation.

Thanks,

William

**Comment by anaphylactic** [comment]
Hi William,

Sorry for the late reply, I'm not keeping a close eye on this these days. The extraction of strongly depleting modules is not implemented in PrisM, it only produces "simply" depleting modules.

Regards,
Ana
